Step2 - 경로 조회 기능 리뷰 요청드립니다.

programmer-sjk opened the pull request:
안녕하세요 동휘님.

이번 미션은 기능 구현전에 mock과 dto를 활용해 인수테스트를 성공시켜야 했는데,
 제가 mock 테스트를 처음 해봐서 조금 2단계 요청이 늦었습니다. 😢 

지난번 `1단계 리뷰`에서 **Custom 예외**와 **public 테스트 함수** 추가했습니다.
동휘님이 말씀하신 `내부 구현에 상관없이 테스트 코드가 작성되어야 한다` 라는 말에 공감했고 하나 배운 느낌이라 저도 기분이 좋았습니다.

2단계 미션 하면서 하나 궁금한 사항이 있습니다.
`PathFinder` 도메인의 최단 경로를 구하는 `shortestPath` 함수인데요

현재 구조는 `shortestPath 함수호출 -> PathResponse` 가 반환이 됩니다. 그래서 서비스에서 바로 return 하는 구조입니다.
이렇게 도메인에서 dto를 반환하는게 혹시 안좋은 방법일까..?

가령 아래와 같은 구조로 만들었어야 했을까 물어보고 싶었습니다.
`shortestPath 함수호출 -> ShortestPath 혹은 Path 라는 response dto 반환 -> 서비스 layer에서 PathResponse로 변환해서 반환`

그럼 2단계 미션 리뷰 부탁드립니다. 🙏

programmer-sjk committed:
refactor: Custom 예외 적용

programmer-sjk committed:
test: Line의 deleteSection 함수 테스트 추가

programmer-sjk committed:
docs(guide): README.md 요구사항 및 기능 구현 목록 작성

programmer-sjk committed:
test(acceptance): 인수 테스트 작성

programmer-sjk committed:
test(acceptance): mock과 dto를 활용해 인수 테스트 성공하도록 수정

programmer-sjk committed:
test(fixture): Line, Section test fixture 추가

programmer-sjk committed:
test(service): mockito 활용한 PathService 테스트 추가

programmer-sjk committed:
feat(service): PathService 추가

programmer-sjk committed:
test(domain): PathFinder 도메인에 대한 테스트 추가

programmer-sjk committed:
test(domain): Lines 일급 컬렉션에 대한 테스트 추가

programmer-sjk committed:
feat(domain): PathFinder 도메인 추가

programmer-sjk committed:
feat(domain): Lines 일급 컬렉션 및 변경사항 추가

programmer-sjk committed:
refactor: 기존 String 에러메시지 -> ErrorCode Enum으로 관리

programmer-sjk committed:
test(acceptance): mock 제거하고 성공하도록 수정

programmer-sjk committed:
refactor: 안쓰는 import 문 제거

programmer-sjk committed:
refactor: 변수명 및 메시지 수정

dhmin5693 commented on src/test/java/nextstep/subway/path/acceptance/PathAcceptanceTest.java:
```diff
@@ -0,0 +1,141 @@
+
+    @DisplayName("지하철 역 사이의 최단 경로를 조회한다")
+    @Test
+    void findShortestPath() {
```
이 테스트가 실패하네요.
로그가 그 이유를 잘 알려주고 있으니 확인해보세요.

만약 테스트가 성공한다면 실행 환경이 gradle로 되어 있을 수 있는데 intellij로 변경하시면 됩니다.

dhmin5693 commented on src/main/java/nextstep/subway/handler/GlobalExceptionHandler.java:
```diff
@@ -1,20 +1,19 @@
 public class GlobalExceptionHandler {
     @ExceptionHandler(IllegalArgumentException.class)
     public ResponseEntity<Void> handleIllegalArgsException() {
         return ResponseEntity.badRequest().build();
     }
 
-    @ExceptionHandler(NoSuchElementException.class)
-    public ResponseEntity<Void> handleNoSuchElementException() {
+    @ExceptionHandler(EntityNotFound.class)
+    public ResponseEntity<Void> handleEntityNotFoundException() {
         return ResponseEntity.badRequest().build();
     }
```
처리 방식이 같으면 하나의 메소드에서 다 처리해도 괜찮지 않을까요?

dhmin5693 commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -57,6 +57,10 @@ private List<Station> sortStations(Map<Station, Station> stations, Station upSta
     }
 
+    public List<Section> get() {
+        return Collections.unmodifiableList(sections);
```
데이터 추출 시 read-only 로 넘겨줘서 내부 데이터가 안전하다고 생각될 수 있는데요,
`Sections` 생성 시 외부에서 받아 온 `List`를 그대로 사용하고 있습니다.
이런 경우에도 예상 외의 데이터 변화가 발생할 수 있어요.
https://tecoble.techcourse.co.kr/post/2021-04-26-defensive-copy-vs-unmodifiable/

dhmin5693 commented on src/main/java/nextstep/subway/path/domain/PathFinder.java:
```diff
@@ -0,0 +1,80 @@
+        initializeEdgeWeight();
+    }
+
+    private void initializeVertax() {
```
```suggestion
    private void initializeVertex() {
```
오타가 있네요

dhmin5693 commented on src/main/java/nextstep/subway/path/domain/PathFinder.java:
```diff
@@ -0,0 +1,80 @@
+import java.util.stream.Collectors;
+
+public class PathFinder {
+    private WeightedMultigraph<Station, DefaultWeightedEdge> graph;
```
final 로 선언되지 않은 이유가 있을까요?
데이터가 변하는 것 같진 않아서요.

dhmin5693 commented:
레이어드 아키텍처의 관점에서 보면 도메인이 DTO를 반환한다 하더라도 큰 문제는 없습니다.
최상위인 표현 영역에서 최하위 영속 영역까지 이어지는 과정에서 역주행만 하지 않는다면요.
DTO는 영역 사이에 데이터를 전달하기 때문에 특정 영역에 속한다고 말하기는 어렵긴 합니다.

지금의 경우엔 표현-서비스-도메인-영속 구조에서 도메인부터 표현까지 그 데이터를 전달하고 있는데요,
`PathResponse` 가 3개 영역에 걸쳐 사용되고 있어서 다소 범위가 넓다고 여겨지긴 하네요.

걸쳐 있는 영역이 좁을 수록 코드 수정 시 영향 범위가 적어집니다.
그렇다고 지금 코드가 안좋다고 말하긴 어렵습니다.
어떤 방법으로 해도 나름의 장단이 있으니 적절하게 대처해주세요.

dhmin5693 reviewed (changes requested):
안녕하세요.

코드를 잘 작성해주셔서 크게 드릴만한 의견은 없네요.
다만 실패하는 테스트가 있어 확인이 필요합니다.
작은 피드백도 몇 개 달아두었으니 같이 확인 부탁드릴게요~

programmer-sjk committed:
refactor: 코드 리뷰 반영

코드 리뷰 반영
- 테스트 실행환경을 IntelliJ로 변경 후 꺠지는 테스트 수정
- ExceptionHandler를 멀티로 처리하도록 수정
- Lines, Sections 생성자에 방어적 복사 추가
- 오타 수정
- 누락한 final 키워드 추가

programmer-sjk commented:
안녕하세요 동휘님. 
말씀대로 실행환경을 `IntelliJ`로 변경했더니 테스트가 깨져서 수정했습니다.
요거에 관련해서는 내일 찾아서 공부해보도록 하겠습니다.

그 외에 리뷰해주신 오타, fianl 누락, 방어적 복사에 대해 수정했습니다.
동휘님이 피드백 주실 때 마다 하나씩 배워가는 느낌입니다. 감사합니다.

dhmin5693 reviewed (approved):
피드백 반영 잘 해주셔서 2단계는 종료하겠습니다.
다음 단계 진행해주세요.
수고하셨습니다 😁
